### PR TITLE
[DeviceSanitizer] refactor options handling and fix use-after-free related problems

### PR DESCRIPTION
**Pull request by yingcong-wu**
This patch:
- refactor options handling.
- for use-after-free, do not try to get allocated/released info when quarantine is not enabled(no such info anyway).
- for findAllocInfoByAddress(), add an assertion to make sure we got the right allocinfo for the address.

SYCLOS PR: https://github.com/intel/llvm/pull/13855

### CODE REVIEW

**Comment by yingcong-wu** [comment]
Hi @wenju-he , @zhaomaosu , @AllanZyne please help reivew this PR first, thank you.

File: source/loader/layers/sanitizer/asan_report.cpp [hunk 1]
```diff
@@ -131,27 +132,35 @@ void ReportUseAfterFree(const DeviceSanitizerReport &Report,
...
-        if (AllocInfo->Context != Context) {
+    if (Options().MaxQuarantineSizeMB > 0) {
+        auto AllocInfoItOp =
+            context.interceptor->findAllocInfoByAddress(Report.Address);
+
+        if (!AllocInfoItOp) {
             context.logger.always("Failed to find which chunck {} is allocated",
                                   (void *)Report.Address);
+        } else {
+            auto &AllocInfo = (*AllocInfoItOp)->second;
```
Comment (cdai2) [inline source/loader/layers/sanitizer/asan_report.cpp:143]: Do you need to assert AllocInfo is not NULL ?

Comment (yingcong-wu) [reply source/loader/layers/sanitizer/asan_report.cpp:143]: AllocInfo will not be NULL here. If not found a right AllocInfo for address, AllocInfoItOp will set to no_value, and will not go to this branch.

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 2]
```diff
@@ -142,65 +143,9 @@ ur_result_t enqueueMemSetShadow(ur_context_handle_t Context,
...
-            std::make_unique<Quarantine>(cl_MaxQuarantineSizeMB * 1024 * 1024);
-    }
-
-    KV = Options->find("detect_locals");
-    if (KV != Options->end()) {
-        auto Value = KV->second.front();
-        cl_DetectLocals = Value == "1" || Value == "true" ? true : false;
+    if (Options().MaxQuarantineSizeMB) {
+        m_Quarantine = std::make_unique<Quarantine>(
+            Options().MaxQuarantineSizeMB * 1024 * 1024);
```
Comment (wenju-he) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:148]:
```suggestion
            static_cast<uint64_t>(Options().MaxQuarantineSizeMB) * 1024 * 1024);
```

**Review by pbalcer — APPROVED** [review]

File: source/loader/layers/sanitizer/asan_options.hpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+#include <algorithm>
+#include <cstring>
+#include <stdexcept>
+
+namespace ur_sanitizer_layer {
+
+struct AsanOptions {
+  public:
+    AsanOptions(AsanOptions &other) = delete;
+    void operator=(const AsanOptions &) = delete;
```
Comment (wenju-he) [inline source/loader/layers/sanitizer/asan_options.hpp:28]:
```suggestion
    AsanOptions operator=(AsanOptions &) = delete;
```
Does copy constructor need to be deleted as well?

Comment (wenju-he) [reply source/loader/layers/sanitizer/asan_options.hpp:28]: oh, I missed the line above

File: source/loader/layers/sanitizer/asan_options.hpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                           [](unsigned char C) { return std::tolower(C); });
+        };
+        auto IsTrue = [&](const std::string &S) {
+            return std::any_of(std::begin(TrueStrings), std::end(TrueStrings),
+                               [&](const char *CS) { return S == CS; });
+        };
+        auto IsFalse = [&](const std::string &S) {
+            return std::any_of(std::begin(FalseStrings), std::end(FalseStrings),
+                               [&](const char *CS) { return S == CS; });
+        };
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_options.hpp:62]: NIT
```suggestion
        auto IsTrue = [TrueStrings](const std::string &S) {
            return std::any_of(std::begin(TrueStrings), std::end(TrueStrings),
                               [&](const char *CS) { return S == CS; });
        };
        auto IsFalse = [FalseStrings](const std::string &S) {
            return std::any_of(std::begin(FalseStrings), std::end(FalseStrings),
                               [&](const char *CS) { return S == CS; });
        };
```

**Review by AllanZyne — APPROVED** [review]

**Comment by yingcong-wu** [comment]
Hi @pbalcer, could we get this PR into the merge pipeline now?

**Pull request merged**